lbcluster/retry_module.go

```diff
@@ -0,0 +1,130 @@
+package lbcluster
+
+import (
+	"fmt"
+	"lb-experts/golbd/logger"
+	"time"
+)
+
+type Retry struct {
+	signal            chan int
+	done              chan bool
+	tick              chan bool
+	currentCount      int
+	maxCount          int
+	retryStarted      bool
+	maxDuration       time.Duration
+	retryDuration     time.Duration
+	prevRetryDuration time.Duration
+	logger            logger.Logger
+}
+
+const defaultMaxDuration = 5 * time.Minute
+
+func NewRetryModule(retryStartDuration time.Duration, logger logger.Logger) *Retry {
+	retry := &Retry{
+		maxCount:          -1,
+		currentCount:      0,
+		maxDuration:       defaultMaxDuration,
+		retryDuration:     retryStartDuration,
+		prevRetryDuration: retryStartDuration,
+		logger:            logger,
+	}
+	return retry
+}
+
+func (r *Retry) SetMaxDuration(maxDuration time.Duration) error {
+	if r.retryStarted {
+		return nil
+	}
+	if maxDuration <= 0 {
+		return fmt.Errorf("duration has to be greater than 0")
+	}
+	r.maxDuration = maxDuration
+	return nil
+}
+
+func (r *Retry) SetMaxCount(maxCount int) error {
+	if r.retryStarted {
+		return nil
+	}
+	if maxCount <= 0 {
+		return fmt.Errorf("max count has to be greater than 0")
+	}
+	r.maxCount = maxCount
+	return nil
+}
+
+func (r *Retry) start() {
+	r.retryStarted = true
+	signal := make(chan int)
+	done := make(chan bool)
+	end := time.Tick(r.maxDuration)
+	r.signal = signal
+	r.done = done
+	go func() {
+		for {
+			select {
+			case <-end:
+				r.done <- true
+				return
+			default:
+				if r.currentCount == r.maxCount {
+					r.done <- true
+					return
+				}
```

Comment:
Is this semantically correct to have return `true` in `r.done` channel even in case of  `r.currentCount == r.maxCount`. Also, should not it return some error in case of all retries have failed?

Reply:
The module never returns the true output to the caller. That is just to terminate the looping. The module does return the error once max retries have reached (Line: 122)/